examples/main/main.cpp

```diff
@@ -269,6 +268,92 @@ void whisper_print_progress_callback(struct whisper_context * /*ctx*/, struct wh
     }
 }
 
+// split UTF-8 string into valid and invalid parts
+// eg. (a = "�123456", result = {"�", "123456", ""})
+// eg. (a = "123456�", result = {"", "123456", "�"})
+// result = {invalid, valid, invalid}
+std::vector<std::string> split_UTF8(const std::string & a) {
+    if (a.empty()) {return {"", "", ""};}
+    std::string str1;
+    std::string str2;
+    std::string str3;
+
+    // forward pass
+    for (int64_t i = 0; i < static_cast<int64_t>(a.length()); i++) {
+        auto value = static_cast<uint8_t>(a[i]);
+        if (value >= 0 && value <= 127 || value >= 192 && value <= 247) {
+            // 1, 2, 3, 4 byte head
+            break;
+        } else if (value >= 128 && value <= 191) {
+            // body byte
+            str1 += a[i];
+        }
+    }
+
+    // backward pass
+    int length = 0;
+    int expect = 0;
+    for (int64_t i = static_cast<int64_t>(a.length()) - 1; i >= 0; i--) {
+        auto value = static_cast<uint8_t>(a[i]);
+        if (value >= 0 && value <= 127) {
+            // 1 byte head
+            expect = 1;
+            length++;
+            break;
+        } else if (value >= 128 && value <= 191){
+            // body byte
+            length++;
+        } else if (value >= 192 && value <= 223){
+            // 2 bytes head
+            expect = 2;
+            length++;
+            break;
+        } else if (value >= 224 && value <= 239){
+            // 3 bytes head
+            expect = 3;
+            length++;
+            break;
+        } else if (value >= 240 && value <= 247){
+            // 4 bytes head
+            expect = 4;
+            length++;
+            break;
+        }
+    }
+    if (expect != length) {
+        str3 = a.substr(a.length() - length, length);
+    }
+
+    str2 = a.substr(str1.length(), a.length() - str3.length());
+
+    if (str1 == str3 && str1.length() + str2.length() + str3.length() > a.length()) {
+        return {str1, str2, ""};
+    }
+    return {str1, str2, str3};
+}
+
+// check whether the start and the end of std::string is encoded in UTF-8
+bool valid_UTF8(const std::string & a) {
+    if (a.empty()) {return true;}
+    auto result = split_UTF8(a);
+    if (result[0].empty() && result[2].empty()) {
+        return true;
+    }
+    return false;
+}
+
+struct UTF8_buf {
+    std::string buffer;     // token buffer (Store incomplete UTF-8 token)
+    float p_sum = 0.0;      // token probability sum
+    int token_c = 0;        // total number of tokens in buffer
+
+    void clear() {
+        buffer = "";
+        p_sum = 0.0;
+        token_c = 0;
+    }
+};
+
```

Comment:
Move these to `common` and rename:

- `utf8_split`
- `utf8_is_valid`
- `utf8_buf`

Reply:
OK